synapse/lib/ast.py

```diff
@@ -3138,7 +3154,63 @@ class AbsProp(Const):
     pass
 
 class Edit(Oper):
-    pass
+
+    def optimize(self):
+
+        if self.optimized:
+            return
+
+        self.gops = [self]
+        self.optimized = True
+        [k.optimize() for k in self.kids]
+
+        # TODO: add deletes to protonode so we can eat those too
+        if isinstance(self, (EditNodeAdd, EditParens)) or not self.isSafeEdit():
```

Comment:
If i'm reading this right, this means an entire edit block is either safe to bulk or not? Perhaps we could detect adjacent edit operations that are safe to bulk rather than being all or nothing?

Reply:
No, this keeps eating operations in the edit block until it reaches an unsafe one. We can also combine EditNodeAdd and EditParens operations but those are a bit more complicated and I haven't added that in yet.